pkg/trustedresources/verify.go

```diff
@@ -223,79 +188,69 @@ func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.
 			passVerification = true
 			break
+		} else {
+			// FixMe: changing %v to %w breaks integration tests.
+			warn := fmt.Errorf("%w:%v", ErrResourceVerificationFailed, err.Error())
+			logger.Warnf(warn.Error())
 		}
 	}
 	return passVerification
 }
 
-// verifyInterface get the checksum of json marshalled object and verify it.
-func verifyInterface(obj interface{}, verifier signature.Verifier, signature []byte) error {
-	ts, err := json.Marshal(obj)
-	if err != nil {
-		return fmt.Errorf("failed to marshal the object: %w", err)
-	}
-
-	h := sha256.New()
-	h.Write(ts)
-
-	if err := verifier.VerifySignature(bytes.NewReader(signature), bytes.NewReader(h.Sum(nil))); err != nil {
-		// FixMe: changing %v to %w breaks integration tests.
-		return fmt.Errorf("%w:%v", ErrResourceVerificationFailed, err.Error())
-	}
-
-	return nil
-}
-
-// prepareObjectMeta will remove annotations not configured from user side -- "kubectl-client-side-apply" and "kubectl.kubernetes.io/last-applied-configuration"
-// (added when an object is created with `kubectl apply`) to avoid verification failure and extract the signature.
-// Returns a copy of the input object metadata with the annotations removed and the object's signature,
-// if it is present in the metadata.
+// extractSignature extracts the signature if it is present in the metadata.
 // Returns a non-nil error if the signature cannot be decoded.
-func prepareObjectMeta(in metav1.Object) (metav1.ObjectMeta, []byte, error) {
-	out := metav1.ObjectMeta{}
-
-	// exclude the fields populated by system.
-	out.Name = in.GetName()
-	out.GenerateName = in.GetGenerateName()
-	out.Namespace = in.GetNamespace()
-
-	if in.GetLabels() != nil {
-		out.Labels = make(map[string]string)
-		for k, v := range in.GetLabels() {
-			out.Labels[k] = v
-		}
-	}
-
-	out.Annotations = make(map[string]string)
-	for k, v := range in.GetAnnotations() {
-		out.Annotations[k] = v
-	}
-
-	// exclude the annotations added by other components
-	// Task annotations are unlikely to be changed, we need to make sure other components
-	// like resolver doesn't modify the annotations, otherwise the verification will fail
-	delete(out.Annotations, "kubectl-client-side-apply")
-	delete(out.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
-
+func extractSignature(in metav1.Object) ([]byte, error) {
 	// signature should be contained in annotation
 	sig, ok := in.GetAnnotations()[SignatureAnnotation]
 	if !ok {
-		return out, nil, nil
+		return nil, nil
 	}
 	// extract signature
 	signature, err := base64.StdEncoding.DecodeString(sig)
 	if err != nil {
-		return out, nil, err
+		return nil, err
 	}
-	delete(out.Annotations, SignatureAnnotation)
+	return signature, nil
+}
 
-	return out, signature, nil
+// getChecksum gets the sha256 checksum of the resource.
+// Returns a non-nil error if the checksum cannot be computed or the resource is of unknown type.
+func getChecksum(resource metav1.Object) ([]byte, error) {
```

Comment:
This is okay, but you could also avoid this by plumbing through an interface that includes the object + checksum interface.

```go
type Hashable interface {
	Checksum() ([]byte, error)
}

type HashableObject interface
	metav1.Object
	Hashable
}
```

(this is probably going to require some refactoring, so feel free to punt this to another pr)

Reply:
You could also type assert on a standalone interface to avoid the type switch:

```go
type Hashable interface {
	Checksum() ([]byte, error)
}

func getChecksum(resource metav1.Object) ([]byte, error) {
	h, ok := o.(Hashable)
	if !ok {
		return nil, fmt.Errorf(...)
	}
	return h.Checksum()
}
```

Reply:
This really helps! thanks